Update Clarity VSCode readme for 1.0 release (#571)
hirosystems/clarinet: components/clarity-vscode/README.md

# Clarity for Visual Studio Code

Clarity is a **decidable** smart contract language that optimizes for predictability and security. Developed by Hiro, Clarity enables smart contract developers to encode essential business logic on a blockchain.

> A programming language is referred to as **decidable** if you can know with certainty, from the code itself, what the program will do. Clarity is was designed to be intentionally Turing incomplete, thereby avoiding the problem of `Turing complexity` (the number of steps it takes for a turing machine to run an algorithm on a given input size). This allows for complete static analysis of the entire call graph of a given smart contract, while Hiro's support for types and type checkers can eliminate whole classes of bugs like unintended casts, reentrancy bugs, and reads of uninitialized values.

Reviewer: I don't think it's necessary to go into too much details about what Clarity is, at least not in the intro of the readme.

Instead, I think some popular extensions tend to list al the implemented features, allowing developers to quickly understand what to expect from the extension

Reviewer 2: +1 to what Hugo says.

@trichards-hiro: This content, in particular, if it can speak to our extension and its value prop, that'd help the readers tremendously. Clarity, as a language, can be ignored entirely from here.

Author: removed the unnecessary Clarity section from intro

> The Language Server Protocol (LSP) defines the protocol used between an editor or IDE, and a language server that provides language features such as auto-complete, go-to-definition, find-all-references, and others.

Reviewer: We probably don't need to get into details about the LSP is

Reviewer 2: Agreed.

Author: removed from intro section

This Clarity-LSP project aims at leveraging the decidability quality of Clarity and the LSP to provide insights about your code, without publishing your smart contracts to a blockchain, while also enabling you to run the Clarity for Visual Studio Code extenstion in-browser.

> ![screenshot](images/screenshot.png)

Reviewer: I'll add the image 👌

Author: Thank you so much for the help.

## Clarity for Visual Studio Installation

> You can install the latest release of the plugin from the [marketplace](https://marketplace.visualstudio.com/items?itemName=hirosystems.clarity-lsp).

Reviewer: ```suggestion
You can install the latest release of the extension directly from VS Code or from the [marketplace](https://marketplace.visualstudio.com/items?itemName=hirosystems.clarity-lsp).
```

The Debugger is not implemented yet on the browser.
![screenshot](images/screenshot.png)

## Debugging
## Clarity for Visual Studio Installation

The debugger included in this extension allows to run Clarity code line by line which is a powerful tool to understand what precisely happens. Read this [blog post](https://www.hiro.so/blog/how-to-debug-your-smart-contracts-with-clarinet) to know more about Clarity debugging.
You can install the latest release of the plugin from the [marketplace](https://marketplace.visualstudio.com/items?itemName=hirosystems.clarity-lsp).
## Features

In order to use the Debugger (DAP), this extension relies on a local installation of Clarinet. To install Clarinet, please follow the instructions [here](https://github.com/hirosystems/clarinet#installation).
The features described below are available when you install the Clarity for Visual Studio plugin.

---

> ## Contributes

Reviewer: I think we should keep this "Contributing" part

Author: Added this back into the README.

> When you select a function, the extension adds the necessary parentheses around it and puts placeholders in the arguments of the function.

Reviewer 2: ```suggestion
For instance, when you select a function, the extension adds the necessary parentheses around it and puts placeholders in the arguments of the function.
```

> ### Check Contract on Save and Display Errors Inline

Reviewer: ```suggestion
### Check Contract on Save and Display Errors
```

Reviewer 2: Too many things on this title. What about:

```suggestion
### Analyze Contract Correctness on Save 
```

![display errors gif](images/errors.gif)

> ### VS-Code Support

Reviewer: ```suggestion
### VS Code for the Web Support
```

> Visual Studio Code is natively supported, but this extension provides additional support for [vscode.dev](https://vscode.dev/), [github.dev](https://github.dev/github/dev) and similar tools.

Reviewer: ```suggestion
This extension works in VS Code on Desktop along with support for [vscode.dev](https://vscode.dev/), [github.dev](https://github.dev/github/dev)
```

Not sure how the phrase this, the point is just to say that the extension is what Microsoft refers as "universal", it works on Desktop and Web (but "natively supported" didn't make much sense to me)

> For more information, please see the [Clarinet GitHub VS Code documentation page](https://github.com/hirosystems/clarinet/blob/develop/components/clarity-vscode/README.md)

Reviewer: This is a link to this page

Author: removed unnecessary link here

### Auto Complete User-Defined Functions

> This feature is similar to the auto-complete feature, but does not provide documentation for functions when writing a contract.

Reviewer: I don't think we should mention "it doesn't provide documentation" for user defined function, since user defined function don't really  have documentation in the first place.

Reviewer: Maybe we can remove it and only change the first feature ("Autocomplete native functions"), with "Autocomplete functions" (native and user defined)

Author: fixed

> ### Resolve Contract-Call Targeting Local Contracts

Reviewer: ```suggestion

### Resolve Contract-Call Targeting Local Contracts
```

This feature is similar to the auto-complete feature, but enables auto-completion over multiple files.

> ### Support for Multiple Errors

Reviewer: I think we can remove this bit.
It's expected from developer when seeing the feature "Check contract on save and display errors"

Author: fixed

Although not considered a native feature for the Visual Studio Code extension, support for multiple errors is now supported in this version.

![multiple error support](images/multicontract.gif)
### Support for Traits

> When a contract implements a trait (such as the NFT of FT trait – SIPs 009 and 010), the extensions will show and errors if the trait implementation is not satisfied (for example, if the trait expects a function which is not implemented, or if the function signature does not match the trait definition).

Reviewer: ```suggestion
When a contract implements a trait (such as the NFT of FT trait – SIPs 009 and 010), the extensions will detect errors if the trait implementation is not satisfied (for example, if the trait expects a function which is not implemented, or if the function signature does not match the trait definition).
```

> ### Documentation

Reviewer: ```suggestion

### Documentation
```

This feature extends the capabilities of the native auto-complete function by providing documentation for a function when writing a smart contract.

> ### Debugger

Reviewer: I would put it higher in the list importance.
Also, we need to mention that in only works on Desktop (at least for now) in that it requires a local installation of clarinet (cf [here](https://github.com/hirosystems/clarinet/pull/571/files#diff-48b9d6c32006f08d5050dbfb8dd4fe9d13005db94738ce97509b0e96c203f7f6L17))

Reviewer 2: Lets, please put the local installation of Clarinet wrapped inside a NOTE or with BOLD formatting so it is front and center and folks know that it is a prerequisite.

Author: I added note in bold stating that this requires a local installation of Clarinet. I also moved this section up in the list of features so it is called out earlier.

### Structure

The LSP has two main parts: the client and the server.
This two part will run different environments:
- VSCode Web:  for Web Workers' environment

> - VSCode Desktop: for Node.js environment```

Reviewer 2: Something is odd here, formatting-wise.

Reviewer: Indeed, something went wrong with the github suggestion I'll clean it
